src/expr/proof_node_manager.cpp

```diff
@@ -83,30 +91,61 @@ std::shared_ptr<ProofNode> ProofNodeManager::mkScope(
     // otherwise it may be due to symmetry?
     Node aeqSym = CDProof::getSymmFact(a);
     Trace("pnm-scope") << "  - try sym " << aeqSym << "\n";
-    if (!aeqSym.isNull())
+    Node aMatch;
+    if (!aeqSym.isNull() && ac.count(aeqSym))
     {
-      if (ac.count(aeqSym))
+      Trace("pnm-scope") << "- reorient assumption " << aeqSym << " via " << a
+                         << " for " << fa.second.size() << " proof nodes"
+                         << std::endl;
+      aMatch = aeqSym;
+    }
+    else
+    {
+      // Otherwise, may be derivable by rewriting. Note this is used in
+      // ensuring that proofs from the proof equality engine that involve
+      // equality with Boolean constants are closed.
+      if (!computedAcr)
       {
-        Trace("pnm-scope") << "- reorient assumption " << aeqSym << " via " << a
-                           << " for " << fa.second.size() << " proof nodes"
-                           << std::endl;
-        std::shared_ptr<ProofNode> pfaa = mkAssume(aeqSym);
-        for (ProofNode* pfs : fa.second)
+        computedAcr = true;
+        for (const Node& acc : ac)
         {
-          Assert(pfs->getResult() == a);
-          // must correct the orientation on this leaf
-          std::vector<std::shared_ptr<ProofNode>> children;
-          children.push_back(pfaa);
-          std::vector<Node> args;
-          args.push_back(a);
-          updateNode(pfs, PfRule::MACRO_SR_PRED_TRANSFORM, children, args);
+          Node accr = theory::Rewriter::rewrite(acc);
+          if (accr != acc)
+          {
+            acr[accr] = acc;
+          }
         }
-        Trace("pnm-scope") << "...finished" << std::endl;
-        acu.insert(aeqSym);
-        continue;
       }
+      Node ar = theory::Rewriter::rewrite(a);
+      std::unordered_map<Node, Node, NodeHashFunction>::iterator itr =
+          acr.find(ar);
+      if (itr != acr.end())
+      {
+        aMatch = itr->second;
+      }
+    }
+
+    // if we found a match either by symmetry or rewriting, then we connect
+    // the assumption here.
```

Comment:
Do I understand correctly that `MACRO_SR_PRED_TRANSFORM` steps are being added for symmetry as well? I'd rather have it special cased.

Reply:
No, we are applying symmetry whenever possible, and only resort to rewriting when this fails.

Reply:
But the update to the proof node, when the assumption changed due to symmetry, is also represented as a `MACRO_SR_PRED_TRANSFORM` step, no?

Reply:
Ah, so yeah, I am preserving the old behavior here, which indeed used MACRO_SR_PRED_TRANSFORM.  You have to be a bit careful here since SYMM has special status in CDProof.

I think we should do an optimization pass later where we eliminate these when possible, perhaps a separate PR.

Reply:
Probably we could defer this to "global optimizations of proof nodes". Perhaps to be done before we expand the coarse grained steps.